backend/ethereum/channel/subscription_test.go

```diff
@@ -0,0 +1,113 @@
+// Copyright 2021 - See NOTICE file for copyright holders.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package channel_test
+
+import (
+	"context"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	"perun.network/go-perun/backend/ethereum/channel/test"
+	"perun.network/go-perun/channel"
+	channeltest "perun.network/go-perun/channel/test"
+	pkgtest "perun.network/go-perun/pkg/test"
+)
+
+func TestSubscription(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	rng := pkgtest.Prng(t)
+	numParts := 2 + rng.Intn(maxNumParts-2)
+	setup := test.NewSetup(t, rng, numParts, blockInterval)
+	adjSetup := newAdjudicatorSetup(setup)
+	adj := adjSetup.Adjudicator()
+
+	opts := []channeltest.RandomOpt{channeltest.WithoutApp(), channeltest.WithIsFinal(false)}
+	req, subChannels := adjSetup.newAdjudicatorReq(ctx, rng, opts...), []channel.SignedState{}
+	sub, err := adj.Subscribe(ctx, req.Params.ID())
+	require.NoError(t, err, "subscribing")
+
+	subSetup := newSubscriptionSetup(sub, adj, req, subChannels)
+	channeltest.TestSubscription(ctx, t, subSetup)
+}
+
+type subscriptionSetup struct {
+	adj         channel.Adjudicator
+	sub         *testSubscription
+	req         *channel.AdjudicatorReq
+	subChannels []channel.SignedState
+}
+
+func newSubscriptionSetup(
+	sub channel.AdjudicatorSubscription,
+	adj channel.Adjudicator,
+	req *channel.AdjudicatorReq,
+	subChannels []channel.SignedState,
+) *subscriptionSetup {
+	return &subscriptionSetup{
+		sub: &testSubscription{
+			emitProgressed:          false,
+			req:                     req,
+			AdjudicatorSubscription: sub,
+		},
+		adj:         adj,
+		req:         req,
+		subChannels: subChannels,
+	}
+}
+
+type testSubscription struct {
+	channel.AdjudicatorSubscription
+	emitProgressed bool
+	req            *channel.AdjudicatorReq
+}
+
+// Next emulates app channel functionality because app channels are not supported yet.
+func (s *testSubscription) Next() channel.AdjudicatorEvent {
+	if s.emitProgressed {
```

Comment:
This assumes that there is only one progression, right?

Reply:
Why do you think it would not work for several progressions? I don't see it.